File: src/org/elixir_lang/psi/impl/QuotableArgumentsImpl.java
package org.elixir_lang.psi.impl;

import com.ericsson.otp.erlang.OtpErlangList;
import com.ericsson.otp.erlang.OtpErlangObject;
import com.ericsson.otp.erlang.OtpErlangTuple;
import com.intellij.psi.PsiElement;
import org.elixir_lang.psi.*;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.elixir_lang.psi.impl.ElixirPsiImplUtil.DO;
import static org.elixir_lang.psi.impl.QuotableImpl.NIL;

public class QuotableArgumentsImpl {
    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(@NotNull final Arguments arguments) {
        PsiElement[] unquotedArguments = arguments.arguments();
        OtpErlangObject[] quotedArguments = new OtpErlangObject[unquotedArguments.length];

        for (int i = 0; i < unquotedArguments.length; i++) {
            Quotable quotableArgument = (Quotable) unquotedArguments[i];
            quotedArguments[i] = quotableArgument.quote();
        }

        return quotedArguments;
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(@NotNull final ElixirBlockList blockList) {
        List<ElixirBlockItem> blockItemList = blockList.getBlockItemList();
        OtpErlangObject[] quotedBlockItems = new OtpErlangObject[blockItemList.size()];

        int i = 0;
        for (ElixirBlockItem blockItem : blockItemList) {
            quotedBlockItems[i++] = blockItem.quote();
        }

        return quotedBlockItems;
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(ElixirUnqualifiedNoParenthesesManyArgumentsCall unqualifiedNoParenthesesManyArgumentsCall) {
        PsiElement[] primaryArguments = unqualifiedNoParenthesesManyArgumentsCall.primaryArguments();
        OtpErlangObject[] quotedArguments = new OtpErlangObject[primaryArguments.length];

        for (int i = 0; i< primaryArguments.length; i++) {
            Quotable quotablePrimaryArgument = (Quotable) primaryArguments[i];
            quotedArguments[i] = quotablePrimaryArgument.quote();
        }

        return quotedArguments;
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(@NotNull final ElixirDoBlock doBlock) {
        ElixirStab stab = doBlock.getStab();
        OtpErlangObject doValue = NIL;
        OtpErlangObject[] quotedKeywordPairs;

        if (stab != null) {
            doValue = stab.quote();
        }

        OtpErlangTuple quotedDoKeywordPair = new OtpErlangTuple(
                new OtpErlangObject[]{
                        DO,
                        doValue
                }
        );

        ElixirBlockList blockList = doBlock.getBlockList();

        if (blockList != null) {
            OtpErlangObject[] blockListQuotedArguments = blockList.quoteArguments();

            quotedKeywordPairs = new OtpErlangObject[1 + blockListQuotedArguments.length];

            int i = 0;
            quotedKeywordPairs[i++] = quotedDoKeywordPair;
            System.arraycopy(blockListQuotedArguments, 0, quotedKeywordPairs, i, blockListQuotedArguments.length);
        } else {
            quotedKeywordPairs = new OtpErlangObject[]{
                    quotedDoKeywordPair
            };
        }

        return new OtpErlangObject[]{
                new OtpErlangList(quotedKeywordPairs)
        };
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(@NotNull final ElixirMapConstructionArguments mapConstructionArguments) {
        PsiElement[] arguments = mapConstructionArguments.arguments();
        List<OtpErlangObject> quotedArgumentList = new ArrayList<>();

        for (PsiElement argument : arguments) {
            Quotable quotableArgument = (Quotable) argument;
            OtpErlangList quotedArgument = (OtpErlangList) quotableArgument.quote();

            quotedArgumentList.addAll(Arrays.asList(quotedArgument.elements()));
        }

        OtpErlangObject[] quotedArguments = new OtpErlangObject[quotedArgumentList.size()];

        return quotedArgumentList.toArray(quotedArguments);
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(@NotNull final ElixirNoParenthesesArguments noParenthesesArguments) {
        PsiElement[] children = noParenthesesArguments.getChildren();

        assert children.length == 1;

        QuotableArguments quotableArguments = (QuotableArguments) children[0];
        return quotableArguments.quoteArguments();
    }

    @Contract(pure = true)
    @NotNull
    public static OtpErlangObject[] quoteArguments(ElixirParenthesesArguments parenthesesArguments) {
        PsiElement[] arguments = parenthesesArguments.arguments();

        OtpErlangObject[] quotedArguments = new OtpErlangObject[arguments.length];

        for (int i = 0; i < arguments.length; i++) {
            Quotable quotableChild = (Quotable) arguments[i];
            OtpErlangObject quotedChild = quotableChild.quote();
            quotedArguments[i] = quotedChild;
        }

        return quotedArguments;
    }
}

File: src/org/elixir_lang/psi/impl/QuotableArgumentsImpl.kt
package org.elixir_lang.psi.impl

import com.ericsson.otp.erlang.OtpErlangAtom
import com.ericsson.otp.erlang.OtpErlangList
import com.ericsson.otp.erlang.OtpErlangObject
import com.ericsson.otp.erlang.OtpErlangTuple
import org.elixir_lang.psi.*
import org.elixir_lang.psi.impl.QuotableImpl.NIL
import org.jetbrains.annotations.Contract

object QuotableArgumentsImpl {
    val DO = OtpErlangAtom("do")

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(arguments: Arguments): Array<OtpErlangObject> =
            arguments
                    .arguments()
                    .map { it as Quotable }
                    .map(Quotable::quote)
                    .toTypedArray()

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(blockList: ElixirBlockList): Array<OtpErlangObject> =
            blockList
                    .blockItemList
                    .map(ElixirBlockItem::quote)
                    .toTypedArray()

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(
            unqualifiedNoParenthesesManyArgumentsCall: ElixirUnqualifiedNoParenthesesManyArgumentsCall
    ): Array<OtpErlangObject> =
            unqualifiedNoParenthesesManyArgumentsCall
                    .primaryArguments()
                    .map { it as Quotable }
                    .map(Quotable::quote)
                    .toTypedArray()

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(doBlock: ElixirDoBlock): Array<OtpErlangObject> {
        val doValue = doBlock.stab?.quote() ?: NIL

        val quotedKeywordPairListPrefix = arrayOf(DO, doValue).let(::OtpErlangTuple).let { listOf(it) }
        val quotedKeywordPairListSuffix = doBlock.blockList?.quoteArguments()?.toList() ?: emptyList()
        val quotedKeywordPairList = quotedKeywordPairListPrefix + quotedKeywordPairListSuffix

        return quotedKeywordPairList.toTypedArray().let(::OtpErlangList).let { arrayOf(it) }
    }

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(mapConstructionArguments: ElixirMapConstructionArguments): Array<OtpErlangObject> =
            mapConstructionArguments
                    .arguments()
                    .map { it as Quotable }
                    .flatMap {
                        it
                                .quote()
                                .let { it as OtpErlangList }
                                .elements()
                                .asList()
                    }.toTypedArray()

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(noParenthesesArguments: ElixirNoParenthesesArguments): Array<OtpErlangObject> =
            noParenthesesArguments
                    .children
                    .single()
                    .let { it as QuotableArguments }
                    .quoteArguments()

    @Contract(pure = true)
    @JvmStatic
    fun quoteArguments(parenthesesArguments: ElixirParenthesesArguments): Array<OtpErlangObject> =
            parenthesesArguments
                    .arguments()
                    .map { it as Quotable }
                    .map(Quotable::quote)
                    .toTypedArray()
}
